### PR TITLE
added GitHub flavored  checklist(task list) markdown

### PR DESCRIPTION
**Pull request by keefo**
* added GitHub flavored  checklist markdown (https://help.github.com/articles/basic-writing-and-formatting-syntax/#task-lists)
* added unit tests.
* fix a MMAssertExtendedMarkdownEqualsHTML nil NSXMLDocument *expected issue.
* move MMAssertExtendedMarkdownEqualsHTML to MMTestCase.h file.

### CODE REVIEW

File: Source/MMMarkdown.h [hunk 1]
```diff
@@ -45,8 +45,9 @@ typedef NS_OPTIONS(NSUInteger, MMMarkdownExtensions)
 //    MMMarkdownExtensionsTableCaptions       = 1 << 7,
     MMMarkdownExtensionsTables              = 1 << 8,
     MMMarkdownExtensionsUnderscoresInWords  = 1 << 9,
+    MMMarkdownExtensionsChecklist           = 1 << 10,  // https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments
```
Comment (mdiep) [inline Source/MMMarkdown.h]: I'd call this `MMMarkdownExtensionsTaskList` to match the GitHub terminology. (And rename the other constants similarly.)

File: Source/MMParser.m [hunk 2]
```diff
@@ -710,6 +726,345 @@ - (MMElement *)_parseHorizontalRuleWithScanner:(MMScanner *)scanner
...
+ - [ ]
+ - [x]
+ - [X]
+ + [ ]
+ * [ ]
+ 1. [ ]
+ 
+ @param scanner MMScanner
+ @return ture if the check list marker exist, false otherwise.
+ */
```
Comment (mdiep) [inline Source/MMParser.m]: This should be formatted to match the existing docstrings in the project.

Also, _true_ has been misspelled.

File: Source/MMParser.m [hunk 3]
```diff
@@ -710,6 +726,345 @@ - (MMElement *)_parseHorizontalRuleWithScanner:(MMScanner *)scanner
...
+                    nextChar = scanner.nextCharacter;
+                    if (nextChar == ' ' || nextChar == 'X' || nextChar == 'x')
+                    {
+                        BOOL foundx = nextChar == 'X' || nextChar == 'x';
+                        [scanner advance];
+                        if (scanner.nextCharacter == ']')
+                        {
+                            [scanner advance];
+                            if (scanner.nextCharacter == ' ')
+                            {
```
Comment (mdiep) [inline Source/MMParser.m]: This should all use early returns to match the existing code.

File: Source/MMParser.m [hunk 4]
```diff
@@ -710,6 +726,345 @@ - (MMElement *)_parseHorizontalRuleWithScanner:(MMScanner *)scanner
...
+ * [ ]
+ 1. [ ]
+ 
+ @param scanner MMScanner
+ @return ture if the check list marker exist, false otherwise.
+ */
+- (BOOL)_parseChecklistMarkerWithScanner:(MMScanner *)scanner listType:(MMElementType*)type expectedListType:(MMElementType)expectedListType checked:(nullable BOOL*)checked
+{
+    // checking
+    if(expectedListType == MMElementTypeNone || expectedListType == MMElementTypeBulletedChecklist)
```
Comment (mdiep) [inline Source/MMParser.m]: space missing here after `if`

File: Source/MMParser.m [hunk 5]
```diff
@@ -710,6 +726,345 @@ - (MMElement *)_parseHorizontalRuleWithScanner:(MMScanner *)scanner
...
+ + [ ]
+ * [ ]
+ 1. [ ]
+ 
+ @param scanner MMScanner
+ @return ture if the check list marker exist, false otherwise.
+ */
+- (BOOL)_parseChecklistMarkerWithScanner:(MMScanner *)scanner listType:(MMElementType*)type expectedListType:(MMElementType)expectedListType checked:(nullable BOOL*)checked
+{
+    // checking
```
Comment (mdiep) [inline Source/MMParser.m]: This comment doesn't really add anything.

File: Source/MMParser.m [hunk 6]
```diff
@@ -710,6 +726,345 @@ - (MMElement *)_parseHorizontalRuleWithScanner:(MMScanner *)scanner
...
+                    }
+                }
+            }
+        }
+        [scanner commitTransaction:NO];
+    }
+ 
+    
+    // checking numbered checklist
+    if(expectedListType == MMElementTypeNone || expectedListType == MMElementTypeNumberedChecklist)
```
Comment (mdiep) [inline Source/MMParser.m]: Space missing here after the `if`

File: Source/MMParser.m (hunk 3 above)
Comment (keefo) [inline Source/MMParser.m]: Thanks for the code review.

After some consideration, I think the current structure is easer to read and understand. 
Plus, _parseListMarkerWithScanner is not strictly using early returns style. =)

**Comment by keefo** [comment]
In some sense, Yes, you are right.  
But I feel that this is more like an extension, so I try to keep it separate and less tangled with the existing list code. If you think about that tasklist is GitHub flavored style, so, we should keep it in this way.

**Comment by mdiep** [comment]
I think that's the right way to go. It will definitely require checks for whether the extension is turned on, but it will require a lot less code.

**Pull request closed without merging**